lib/Sema/CSDiagnostics.h

```diff
@@ -807,20 +807,54 @@ class MissingCallFailure final : public FailureDiagnostic {
   bool diagnoseAsError() override;
 };
 
-class MissingPropertyWrapperUnwrapFailure final : public ContextualFailure {
-  DeclName PropertyName;
+class PropertyWrapperReferenceFailure : public ContextualFailure {
+  VarDecl *Property;
+  Optional<DeclName> MemberName;
+  bool UsingStorageWrapper;
 
+public:
+  PropertyWrapperReferenceFailure(Expr *root, ConstraintSystem &cs,
+                                  VarDecl *property,
+                                  Optional<DeclName> memberName,
+                                  bool usingStorageWrapper, Type base,
+                                  Type wrapper, ConstraintLocator *locator)
+      : ContextualFailure(root, cs, base, wrapper, locator), Property(property),
+        MemberName(memberName), UsingStorageWrapper(usingStorageWrapper) {}
+
+  VarDecl *getProperty() const { return Property; }
+  Optional<DeclName> getMemberName() const { return MemberName; }
+  bool usingStorageWrapper() const { return UsingStorageWrapper; }
```

Comment:
Probably makes sense to add `StringRef getPropertyName() const` here to use in different specific diagnostics.